### PR TITLE
[WIP]Instrumentation, test and fix repo for event lost issues

### PR DESCRIPTION
**Pull request by wenjiaswe**
**What this PR does / why we need it**:
Create a public repo for instrumentation, test and fix contribution for event lost issues

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This issue is only for event lost issues investigation purpose.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by wenjiaswe** [comment]
cc @roycaihw @cheftako @yliaog @janetkuo @jpbetz , I opened this PR so we could all commit to the same branch.

**Comment by wenjiaswe** [comment]
Signed CLA as Google employee

**Comment by wenjiaswe** [comment]
THIS IS ONLY USED TEMPORARILY FOR DEBUG PURPOSE.

@roycaihw @cheftako @jpbetz , I still have a list of TODO as shown in the last paragraph. Please help to see if there is anything missing or other suggestions.

 I have added event logging points in:
1. staging/src/k8s.io/apiserver/pkg/storage/wache_cache.go processEvent, this is where kube-apiserver updates cache with watched events (Record the events coming into apiserver)
2. staging/src/k8s.io/apiserver/pkg/storage/cacher.go sendWatchCacheEvent, this is where watcher processes the event and sends to the result channel  (Record the events going out of apiserver)
3. staging/src/k8s.io/client-go/tools/cache/reflector.go, watchHandler, it gets the events from result channel and watch using the resourceVersion (Record events trace in a couple of different place: 1. when first sync to the latest resource version, 2. apiserver startCaching, 3. shared by clients like scheduler, controller-manager).

At each point, SWAT is used as the label of logging, followed by file name and function name, then event type, meta.name, and event resourceVersion (this is monochronic for each event) are recored. Now each log entry looks like this:

For a pod creation: 
$ kubectl create -f https://github.com/kubernetes/website/blob/master/docs/user-guide/walkthrough/pod-nginx.yaml
Logs in kube-apiserver.log:

SWAT(reflector.go watchHandler) event type, name, resourceVersion: ADDED, nginx, 968  (Cacher calls reflector to get the resource version at the moment of call)
SWAT(watch_cache.go processEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968
SWAT(cacher.go sendWatchCacheEvent) event type, name, resourceVersion: ADDED, nginx, 968 (need to find a better place so no repeating messages)
SWAT(reflector.go watchHandler) event type, name, resourceVersion: ADDED, nginx, 968 (Cacher startCaching)

Logs in kube-scheduler.log:
SWAT(reflector.go watchHandler) event type, name, resourceVersion: ADDED, nginx, 968

Logs in kube-controller-manager.go:
SWAT(reflector.go watchHandler) event type, name, resourceVersion: ADDED, nginx, 968

**TODO**
1. As @cheftako suggested, record what exactly are the events that apiserver got and the events that apiserver sends out, so we know if there is discrepancy there. 
2. A simple comparer that compare the list of resourceVersions at different locations and see if there is any one missing. (No need to wait on this for running test, comparing tool is for easiness of analysis, as long as we have all the data points recorded, we can always look at it or use the comparer whenever available.)
3. Simplify log, now it takes too much memories. Maybe ok for now, depends on tests. If the test contains large clusters and lots of data points the memory issue would be big. 
4. Add data points in scheduler and controller-manager and see how they actually used the events info from reflector.
5. Working on etcd log parser to get the all the actual writing of event out. (Again, tests can be run with the data points as-is now, we don't need to wait for the etcd part, as all the info should be recorded in etcd data folder, as long as we save the file from the test runs, we can always look at it or use the parser whenever available.)

**Comment by wenjiaswe** [comment]
cc @mml

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -392,6 +392,10 @@ loop:
 				continue
 			}
 			newResourceVersion := meta.GetResourceVersion()
+			// SWAT Event lost: print out resourceVersion, EventType and event object name
+			// if event is for a pod or a node
+			fmt.Printf("SWAT,reflector/watchHandler,%s,%s,%s\n",
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: I see reflector being used in multiple components (both in server and clients). Can we add `reflector.name` in the log to identify which component we are looking at?

**Review by roycaihw — COMMENTED** [review]
@wenjiaswe Thanks for the contribution. I left some comments. I understand that you are planning on switching to glog verbose level and some other name for this instrumentation, and also combining the similar logs for `cacher.watchers`.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher.go [hunk 2]
```diff
@@ -888,6 +888,15 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
...
 	}
 
+	// SWAT Event lost: print out resourceVersion, EventType and event object name
+	// if event is for a pod or a node
+	meta, err := meta.Accessor(watchEvent.Object)
+	if err != nil {
+		return
+	}
+	fmt.Printf("SWAT,cacher/sendWatchCacheEvent,%s,%s,%s\n",
+		watchEvent.Type, meta.GetName(), meta.GetResourceVersion())
```
Comment (roycaihw) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: It looks like most (if not all) of the `runtime.Object` carried by these events don't set the `TypeMeta`. As a result `watchEvent.Object.GetObjectKind().GroupVersionKind()` returns empty GVK. Don't know why that comes to happen (maybe apiserver doesn't need to know gvk when it's serving objects from etcd to clients). 

You could use `reflect.TypeOf()` to identify the Go type for now.

Comment (wenjiaswe) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: OK, thanks. I will do that.

**Comment by wenjiaswe** [comment]
@roycaihw new update pushed.

**Pull request closed without merging**